src/test/java/com/liferay/faces/util/cache/CacheTest.java

```diff
@@ -0,0 +1,319 @@
+		}
+
+		for (Thread thread : threads) {
+			thread.start();
+		}
+
+		errorOrFailTestIfNecessary(testErrors, testFailures);
+	}
+
+	private void testConcurrentMaxCacheCapacityCache(final Cache cache, final int maxCacheCapacity,
+		Queue<Throwable> testErrors, Queue<AssertionError> testFailures) throws Exception {
+
+		if (maxCacheCapacity < 10) {
+			throw new IllegalArgumentException("This test must be run with at least 10 values.");
+		}
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+
+					for (int i = 0; i < maxCacheCapacity; i++) {
+						cache.putIfAbsent("key" + i, "value" + i);
+					}
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					Assert.assertNotNull(cache.get("key0"));
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.putIfAbsent("key1", "value1");
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.put("key2", "value2");
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.put("key" + maxCacheCapacity, "value" + maxCacheCapacity);
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					Assert.assertNull(cache.get("key3"));
+				}
+			}.run();
+
+		errorOrFailTestIfNecessary(testErrors, testFailures);
+	}
+
+	private abstract static class TestThreadBase extends Thread {
+
+		// Protected Final Data Members
+		protected final Cache<String, String> cache;
+		protected final Queue<AssertionError> testFailures;
+		protected final Queue<Throwable> testErrors;
+		protected final Integer sleepTimeInMillis;
+
+		public TestThreadBase(Cache<String, String> cache, Queue<AssertionError> testFailures,
+			Queue<Throwable> testErrors) {
+			this(cache, testFailures, testErrors, null);
+		}
+
+		public TestThreadBase(Cache<String, String> cache, Queue<AssertionError> testFailures,
+			Queue<Throwable> testErrors, Integer sleepTimeInMillis) {
+			this.cache = cache;
+			this.testFailures = testFailures;
+			this.testErrors = testErrors;
+			this.sleepTimeInMillis = sleepTimeInMillis;
+		}
+
+		@Override
+		public void run() {
+
+			if (sleepTimeInMillis != null) {
+
+				try {
+					sleep(sleepTimeInMillis);
+				}
+				catch (InterruptedException e) {
+					// do nothing.
+				}
+			}
+
+			try {
+				testCache();
+			}
+			catch (AssertionError e) {
+				testFailures.add(e);
+			}
+			catch (Throwable e) {
+				testErrors.add(e);
+			}
+		}
+
+		protected abstract void testCache();
+	}
+
+	private static final class TestConcurrentCacheThread extends TestThreadBase {
+
+		// Private Final Data Members
+		protected final String key;
```

Comment:
@stiemannkj1 IDE warns that these "protected" data members should be package-private.